lib/ansible/modules/cloud/azure/azure_rm_virtualmachine_facts.py

```diff
@@ -0,0 +1,422 @@
+    pass
+
+from ansible.module_utils.azure_rm_common import AzureRMModuleBase, azure_id_to_dict
+
+
+AZURE_OBJECT_CLASS = 'VirtualMachine'
+
+AZURE_ENUM_MODULES = ['azure.mgmt.compute.models']
+
+
+class AzureRMVirtualMachineFacts(AzureRMModuleBase):
+
+    def __init__(self):
+
+        self.module_arg_spec = dict(
+            resource_group=dict(type='str'),
+            name=dict(type='str'),
+            tags=dict(type='list'),
+            auth_source=dict(type='str', choices=['auto', 'cli', 'env', 'credential_file', 'msi'], default='auto'),
+            cloud_environment=dict(type='str', default='AzureCloud')
+        )
+
+        self.results = dict(
+            changed=False,
+            ansible_facts=dict(azure_virtualmachines=[])
+        )
+
+        self.resource_group = None
+        self.name = None
+        self.tags = None
+
+        super(AzureRMVirtualMachineFacts, self).__init__(self.module_arg_spec,
+                                                         supports_tags=False,
+                                                         facts_module=True)
+
+    def exec_module(self, **kwargs):
+
+        for key in self.module_arg_spec:
+            setattr(self, key, kwargs[key])
+
+        if self.name and not self.resource_group:
+            self.fail("Parameter error: resource group required when filtering by name.")
+        if self.name:
+            self.results['ansible_facts']['azure_virtualmachines'] = self.get_item()
+        else:
+            self.results['ansible_facts']['azure_virtualmachines'] = self.list_items()
+
+        return self.results
+
+    def get_item(self):
+        self.log('Get properties for {0}'.format(self.name))
+        item = None
+        result = []
+
+        try:
+            item = self.compute_client.virtual_machines.get(self.resource_group, self.name)
+        except CloudError:
+            pass
+
+        if item and self.has_tags(item.tags, self.tags):
+            result = [self.serialize_vm(self.get_vm(item.name))]
+
+        return result
+
+    def list_items(self):
+        self.log('List all items')
+        try:
+            items = self.compute_client.virtual_machines.list(self.resource_group)
+        except CloudError as exc:
+            self.fail("Failed to list all items - {0}".format(str(exc)))
+
+        results = []
+        for item in items:
+            if self.has_tags(item.tags, self.tags):
+                results.append(self.serialize_vm(self.get_vm(item.name)))
+        return results
+
+    def get_vm(self, name):
+        '''
+        Get the VM with expanded instanceView
+
+        :return: VirtualMachine object
+        '''
+        try:
+            vm = self.compute_client.virtual_machines.get(self.resource_group, name, expand='instanceview')
+            return vm
+        except Exception as exc:
+            self.fail("Error getting virtual machine {0} - {1}".format(self.name, str(exc)))
+
+    def serialize_vm(self, vm):
```

Comment:
I think you could just use vm.as_dict() instead. or are you actually simplifying the results here?

Reply:
I'm sorry, I didn't get it. Maybe the last commit has the answer.

Reply:
i meant that maybe you don't need serialize_vm function at all, you could just use as_dict() which is available from vm object

Reply:
I used almost the same serialize_vm function of the azure_rm_virtualmachine module. The idea is to return the same dict structure of the azure_rm_virtualmachine module.

Reply:
ah, ok... yes, i think when original azure_rm_virtualmachine was done, as_dict() was not available yet. that's why there's serialize_vm function in there. I will actually try to check how it differs from raw REST API response....

Reply:
The serialize_vm function basically expands the network interfaces, adding useful data like IP addresses.

Reply:
@zikalino Is the PR ready to be approved?